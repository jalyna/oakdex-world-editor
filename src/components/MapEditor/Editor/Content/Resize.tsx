import * as React from 'react'
import styled from 'styled-components'
import { Dispatch } from 'redux'
import { connect } from 'react-redux'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faCaretRight, faCaretDown, faCaretLeft, faCaretUp } from '@fortawesome/free-solid-svg-icons'

import Button from 'shared/Button'

import { Layer } from 'components/MapEditor/reducers/mapData'
import { UPDATE_MAP } from 'components/MapEditor/actionTypes'
import store from 'components/MapEditor/store'

interface ResizeProps {
  changeSize: (direction: string, changeBy: number) => void
}

function mapStateToProps ({ mapData, editorData }: any) {
  return {}
}

function moveLayersByOffset (layers: Layer[], offsetX: number, offsetY: number): Layer[] {
  return layers.slice().map((layer) => {
    let newLayer = { ...layer }
    newLayer.fields = newLayer.fields.slice().map((field) => {
      return {
        ...field,
        x: field.x + offsetX,
        y: field.y + offsetY
      }
    })
    return newLayer
  })
}

function mapDispatchToProps (dispatch: Dispatch) {
  return {
    changeSize: (direction: string, changeBy: number) => {
      let mapData = { ...store.getState().mapData }
      if (direction === 'top' || direction === 'bottom') {
        mapData.height = mapData.height + changeBy
      } else if (direction === 'left' || direction === 'right') {
        mapData.width = mapData.width + changeBy
      }

      let newLayers = mapData.layers.slice()

      if (direction === 'top') {
        newLayers = moveLayersByOffset(newLayers, 0, changeBy)
      } else if (direction === 'left') {
        newLayers = moveLayersByOffset(newLayers, changeBy, 0)
      }

      newLayers = newLayers.map((layer) => {
        let newLayer = { ...layer }
        newLayer.fields = newLayer.fields.slice().filter((field) => {
          return field.x < mapData.width && field.y < mapData.height && field.x >= 0 && field.y >= 0
        })
        return newLayer
      })

      mapData.layers = newLayers

      dispatch({
        type: UPDATE_MAP,
        data: { ...mapData }
      })
    }
  }
}

function Resize ({ changeSize }: ResizeProps) {
  return (
    <React.Fragment>
      <ResizeRight>
        <Button onClick={changeSize.bind(this, 'right', 1)}><FontAwesomeIcon icon={faCaretRight} /></Button>
        <Button onClick={changeSize.bind(this, 'right', -1)}><FontAwesomeIcon icon={faCaretLeft} /></Button>
      </ResizeRight>
      <ResizeLeft>
        <Button onClick={changeSize.bind(this, 'left', 1)}><FontAwesomeIcon icon={faCaretLeft} /></Button>
        <Button onClick={changeSize.bind(this, 'left', -1)}><FontAwesomeIcon icon={faCaretRight} /></Button>
      </ResizeLeft>
      <ResizeTop>
        <Button onClick={changeSize.bind(this, 'top', 1)}><FontAwesomeIcon icon={faCaretUp} /></Button>
        <Button onClick={changeSize.bind(this, 'top', -1)}><FontAwesomeIcon icon={faCaretDown} /></Button>
      </ResizeTop>
      <ResizeBottom>
        <Button onClick={changeSize.bind(this, 'bottom', 1)}><FontAwesomeIcon icon={faCaretDown} /></Button>
        <Button onClick={changeSize.bind(this, 'bottom', -1)}><FontAwesomeIcon icon={faCaretUp} /></Button>
      </ResizeBottom>
    </React.Fragment>
  )
}

const ResizeItem = styled.div`
  position: absolute;
  width: 45px;
  box-sizing: border-box;
  padding: 5px;
  display: flex;

  button {
    margin: 0 3px;
  }
`

const ResizeItemX = styled(ResizeItem)`
  flex-direction: column;
  justify-content: center;
  align-items: stretch;

  button {
    margin: 3px 0;
    flex-grow: 1;
  }
`

const ResizeRight = styled(ResizeItemX)`
  top: 0;
  bottom: 0;
  left: 100%;
`

const ResizeLeft = styled(ResizeItemX)`
  top: 0;
  bottom: 0;
  right: 100%;
`

const ResizeTop = styled(ResizeItem)`
  left: 0;
  right: 0;
  width: 100%;
  bottom: 100%;
`

const ResizeBottom = styled(ResizeItem)`
  left: 0;
  right: 0;
  width: 100%;
  top: 100%;
`

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Resize)
