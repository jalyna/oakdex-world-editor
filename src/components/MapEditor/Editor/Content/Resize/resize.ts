import { Dispatch } from 'redux'

import store from 'components/MapEditor/store'
import { UPDATE_MAP } from 'components/MapEditor/actionTypes'
import { Layer } from 'components/MapEditor/reducers/mapData'

interface MapDataSubset {
  width?: number,
  height?: number,
  layers?: Layer[]
}

function moveLayersByOffset (layers: Layer[], offsetX: number, offsetY: number): Layer[] {
  return layers.slice().map((layer) => {
    let newLayer = { ...layer }
    newLayer.fields = newLayer.fields.slice().map((field) => {
      return {
        ...field,
        x: field.x + offsetX,
        y: field.y + offsetY
      }
    })
    return newLayer
  })
}

function changeDimensions(direction: string, changeBy: number) {
  const mapData = store.getState().mapData
  let data = {
    width: mapData.width,
    height: mapData.height
  } as MapDataSubset

  if (direction === 'top' || direction === 'bottom') {
    data.height = mapData.height + changeBy
  } else if (direction === 'left' || direction === 'right') {
    data.width = mapData.width + changeBy
  }

  return data
}

function removeOutsideTiles (mapData: MapDataSubset, layers: Layer[]): Layer[] {
  return layers.map((layer) => {
    let newLayer = { ...layer }
    newLayer.fields = newLayer.fields.slice().filter((field) => {
      return field.x < mapData.width && field.y < mapData.height && field.x >= 0 && field.y >= 0
    })
    return newLayer
  })
}


export default function (dispatch: Dispatch, direction: string, changeBy: number) {
  const mapData = store.getState().mapData
  let data = changeDimensions(direction, changeBy)

  data.layers = removeOutsideTiles(data, moveLayersByOffset(
    mapData.layers.slice(),
    (direction === 'left' ? changeBy : 0),
    (direction === 'top' ? changeBy : 0)
  ))

  dispatch({
    type: UPDATE_MAP,
    data
  })
}
