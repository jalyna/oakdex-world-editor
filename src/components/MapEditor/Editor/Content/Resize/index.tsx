import * as React from 'react'
import styled, { StyledComponent } from 'styled-components'
import { Dispatch } from 'redux'
import { connect } from 'react-redux'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faCaretRight, faCaretDown, faCaretLeft, faCaretUp, IconDefinition } from '@fortawesome/free-solid-svg-icons'

import Button from 'shared/Button'

import resize from './resize'

interface IconMap {
  [key: string]: IconDefinition
}

interface ContainerMap {
  [key: string]: StyledComponent<"div", any>
}

const ICON_MAP = {
  'right_1': faCaretRight,
  'right_-1': faCaretLeft,
  'left_1': faCaretLeft,
  'left_-1': faCaretRight,
  'top_1': faCaretUp,
  'top_-1': faCaretDown,
  'bottom_1': faCaretDown,
  'bottom_-1': faCaretUp
} as IconMap

const ResizeItem = styled.div`
  position: absolute;
  width: 45px;
  box-sizing: border-box;
  padding: 5px;
  display: flex;

  button {
    margin: 0 3px;
  }
`

const ResizeItemX = styled(ResizeItem)`
  flex-direction: column;
  justify-content: center;
  align-items: stretch;

  button {
    margin: 3px 0;
    flex-grow: 1;
  }
`

const MARGIN = 32

const ResizeRight = styled(ResizeItemX)`
  top: 0;
  bottom: 0;
  left: 100%;
  margin-left: ${MARGIN}px;
`

const ResizeLeft = styled(ResizeItemX)`
  top: 0;
  bottom: 0;
  right: 100%;
  margin-right: ${MARGIN}px;
`

const ResizeTop = styled(ResizeItem)`
  left: 0;
  right: 0;
  width: 100%;
  bottom: 100%;
  margin-bottom: ${MARGIN}px;
`

const ResizeBottom = styled(ResizeItem)`
  left: 0;
  right: 0;
  width: 100%;
  top: 100%;
  margin-top: ${MARGIN}px;
`

const SpecialButton = styled(Button)`
  background-color: rgba(255, 255, 255, 0.5)
`

const CONTAINER_MAP = {
  right: ResizeRight,
  left: ResizeLeft,
  top: ResizeTop,
  bottom: ResizeBottom
} as ContainerMap

interface ResizeProps {
  changeSize: (direction: string, changeBy: number) => void
}

function mapDispatchToProps (dispatch: Dispatch) {
  return {
    changeSize: resize.bind(null, dispatch)
  }
}

function renderButton (changeSize: (direction: string, changeBy: number) => void, direction: string, changeBy: number) {
  return (
    <SpecialButton onClick={changeSize.bind(this, direction, changeBy)}>
      <FontAwesomeIcon icon={ICON_MAP[[direction, changeBy].join('_')]} />
    </SpecialButton>
  )
}

function renderDirection (changeSize: (direction: string, changeBy: number) => void, direction: string) {
  const Container = CONTAINER_MAP[direction]
  return (
    <Container key={direction}>
      {renderButton(changeSize, direction, 1)}
      {renderButton(changeSize, direction, -1)}
    </Container>
  )
}

function Resize ({ changeSize }: ResizeProps) {
  return (
    <React.Fragment>
      {Object.keys(CONTAINER_MAP).map((dir) => {
        return renderDirection(changeSize, dir)
      })}
    </React.Fragment>
  )
}

export default connect(
  () => ({}),
  mapDispatchToProps,
)(Resize)
