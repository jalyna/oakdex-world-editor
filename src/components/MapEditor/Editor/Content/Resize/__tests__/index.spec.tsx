import * as React from 'react'
import configureStore, { MockStoreEnhanced } from 'redux-mock-store'
import { Provider } from 'react-redux'
import { mount } from 'enzyme'
import toJson from 'enzyme-to-json'

import Resize from '..'

const mockStore = configureStore()
let store = null as MockStoreEnhanced | null

describe('<Resize>', () => {
  beforeEach(() => {
    store = mockStore({})
  })

  it('renders correctly', () => {
    const editor = mount(<Provider store={store}><Resize /></Provider>)
    expect(toJson(editor, { noKey: false, mode: 'deep' })).toMatchSnapshot()
  })
})
