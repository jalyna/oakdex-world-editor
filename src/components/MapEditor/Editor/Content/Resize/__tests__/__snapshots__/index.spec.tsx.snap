// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Resize> renders correctly 1`] = `
<Connect(Resize)>
  Array [
    <Styled(Styled(styled.div))
      key="right"
    >
      <div
        className="sc-hrWEMg sc-eTuwsz sc-gwVKww dbEAK"
      >
        <styled.button
          onClick={[Function]}
        >
          <button
            className="sc-bdVaJa dJicAm"
            onClick={[Function]}
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-caret-right fa-w-6 "
              data-icon="caret-right"
              data-prefix="fas"
              focusable="false"
              role="img"
              style={Object {}}
              viewBox="0 0 192 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </button>
        </styled.button>
        <styled.button
          onClick={[Function]}
        >
          <button
            className="sc-bdVaJa dJicAm"
            onClick={[Function]}
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-caret-left fa-w-6 "
              data-icon="caret-left"
              data-prefix="fas"
              focusable="false"
              role="img"
              style={Object {}}
              viewBox="0 0 192 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </button>
        </styled.button>
      </div>
    </Styled(Styled(styled.div))>,
    <Styled(Styled(styled.div))
      key="left"
    >
      <div
        className="sc-hrWEMg sc-eTuwsz sc-hXRMBi bUdirm"
      >
        <styled.button
          onClick={[Function]}
        >
          <button
            className="sc-bdVaJa dJicAm"
            onClick={[Function]}
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-caret-left fa-w-6 "
              data-icon="caret-left"
              data-prefix="fas"
              focusable="false"
              role="img"
              style={Object {}}
              viewBox="0 0 192 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </button>
        </styled.button>
        <styled.button
          onClick={[Function]}
        >
          <button
            className="sc-bdVaJa dJicAm"
            onClick={[Function]}
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-caret-right fa-w-6 "
              data-icon="caret-right"
              data-prefix="fas"
              focusable="false"
              role="img"
              style={Object {}}
              viewBox="0 0 192 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </button>
        </styled.button>
      </div>
    </Styled(Styled(styled.div))>,
    <Styled(styled.div)
      key="top"
    >
      <div
        className="sc-hrWEMg sc-epnACN gaEhFw"
      >
        <styled.button
          onClick={[Function]}
        >
          <button
            className="sc-bdVaJa dJicAm"
            onClick={[Function]}
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-caret-up fa-w-10 "
              data-icon="caret-up"
              data-prefix="fas"
              focusable="false"
              role="img"
              style={Object {}}
              viewBox="0 0 320 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M288.662 352H31.338c-17.818 0-26.741-21.543-14.142-34.142l128.662-128.662c7.81-7.81 20.474-7.81 28.284 0l128.662 128.662c12.6 12.599 3.676 34.142-14.142 34.142z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </button>
        </styled.button>
        <styled.button
          onClick={[Function]}
        >
          <button
            className="sc-bdVaJa dJicAm"
            onClick={[Function]}
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-caret-down fa-w-10 "
              data-icon="caret-down"
              data-prefix="fas"
              focusable="false"
              role="img"
              style={Object {}}
              viewBox="0 0 320 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </button>
        </styled.button>
      </div>
    </Styled(styled.div)>,
    <Styled(styled.div)
      key="bottom"
    >
      <div
        className="sc-hrWEMg sc-iQNlJl gFhoGq"
      >
        <styled.button
          onClick={[Function]}
        >
          <button
            className="sc-bdVaJa dJicAm"
            onClick={[Function]}
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-caret-down fa-w-10 "
              data-icon="caret-down"
              data-prefix="fas"
              focusable="false"
              role="img"
              style={Object {}}
              viewBox="0 0 320 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </button>
        </styled.button>
        <styled.button
          onClick={[Function]}
        >
          <button
            className="sc-bdVaJa dJicAm"
            onClick={[Function]}
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-caret-up fa-w-10 "
              data-icon="caret-up"
              data-prefix="fas"
              focusable="false"
              role="img"
              style={Object {}}
              viewBox="0 0 320 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M288.662 352H31.338c-17.818 0-26.741-21.543-14.142-34.142l128.662-128.662c7.81-7.81 20.474-7.81 28.284 0l128.662 128.662c12.6 12.599 3.676 34.142-14.142 34.142z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </button>
        </styled.button>
      </div>
    </Styled(styled.div)>,
  ]
</Connect(Resize)>
`;
